### ISSUE TITLE
"t_cal" in plot_cumulative_transactions.. what does this do? 

### ISSUE THREAD

**Issue by soliverc** [CamDavidsonPilon/lifetimes#334]
I can't find any info on this parameter in the `plot_cumulative_transactions` function. 

According to the documentation, `t_cal` is  "A marker used to indicate where the vertical line for plotting should be."

OK I can put a vertical red line anywhere on the plot, but what does this do? No matter which value I set as `t_cal` the plot doesn't change. 

What should I do with this parameter? 

Thanks

**Comment by psygo** [comment]
That's just a marker line for the limit *between the calibration (train) and the holdout (test)* periods (the "cal" in `t_cal` is for *calibration*). It's supposed to be in the same position to the parameters you used when separating the data previously, but you can put it anywhere you want.

**Comment by soliverc** [comment]
Ok thanks. Just to clarify the steps I need to take if you don't mind. (Sorry, there aren't a lot of clear tutorials on this library).

I first prepare the data with `summary = summary_data_from_transaction_data(df,etc..)`

Then fit a model to the data. 
```python
bgf = BetaGeoFitter(penalizer_coef=0.001)
bgf.fit(summary ['frequency'], summary ['recency'], summary ['T']) # fit the model
```
Now my model `bgf` is ready to go. We can get expected purchases etc using `bgf.conditional_expected_number_of_purchases_up_to_time`

However, in the next step (I'm following along the documentation) I prepare calibration vs holdout data and now I fit the `bgf` model to that. 

```python
summary_cal_holdout = calibration_and_holdout_data(df, 'id', 'date',
                                        calibration_period_end=split_date)

bgf.fit(summary_cal_holdout['frequency_cal'], summary_cal_holdout['recency_cal'], summary_cal_holdout['T_cal'])
```
It's recommended to do some diagnostic plots at this stage such as `plot_calibration_purchases_vs_holdout_purchases`

And finally the next step is a guide on predicting the probability of your customer being "Alive" with `bgf.predict()`

The documentation uses the `bgf` fitted to the calibration/holdout data for predicting on the original summary data. Should I re-fit `bgf` back to the original data before carrying out these `P-Alive` and `CLV` predictions?

**Comment by psygo** [comment]
> Ok thanks. I'm not sure why we need to put that line there, as we are not using `plot_cumulative_transactions` on the data that was split into calibration/holdout. It's plotted on the full un-split transaction data.

When you study your data, you will eventually use the split data and then use *only* the model that has been fit with the calibration/training data, because the holdout/testing data should be mostly used for control/quality purposes. Then, later, you can check if your model is doing well on both the training and test data by plotting the `plot_cumulative_transactions` and `plot_incremental_transactions` graphs. Having the `t_cal` marker will help you visually differentiate where those two (sub)datasets are located on the graph.

> However, in the next step (I'm following along the documentation) I prepare calibration vs holdout data and now I fit the bgf model to that.

You do *not* need to fit your model first on the whole data and then on the train/test data. You can go directly to the train/test data if you wish to. Fitting to the whole dataset is supposed to only serve as a guideline to the final results, but, in the end, you will need to have a training and a control (test) dataset in order to statistically validate your results.

> It's recommended to do some diagnostic plots at this stage such as `plot_calibration_purchases_vs_holdout_purchases`

That's one of the validation plots, not the only one. It's important to use everything that's available to you. Many of the plots only show that your model *does not suck*, and *not* that it is good.

> The documentation uses the `bgf` fitted to the calibration/holdout data for predicting on the original summary data. Should I re-fit `bgf` back to the original data before carrying out these `P-Alive` and `CLV` predictions?

Rigorously, you shouldn't refit your model to the whole dataset because everything it has "learned" to get its parameters and to validate its assumptions comes from the training set and not the complete data. Using the model you obtained from your calibration/holdout study on the whole data is, thus, valid, but not refitting it to the whole data and then using it on the same whole set &mdash; that's also basically like cheating on a school test: once you know the content of the test, it becomes much easier to answer the questions.

**Comment by soliverc** [comment]
> That's one of the validation plots, not the only one. It's important to use everything that's available to you. Many of the plots only show that your model does not suck, and not that it is good.

Which ones should I use to check if the model _is good?_ 

Also, is there a way to check CLV accuracy similar to `plot_calibration_purchases_vs_holdout_purchases`?

**Comment by psygo** [comment]
> Which ones should I use to check if the model *is good?*

All of the available ones.

> Also, is there a way to check CLV accuracy similar to `plot_calibration_purchases_vs_holdout_purchases`? 

I think you could use the sum of each customer's past purchases as an estimate of their CLV and check if `probability of a customer being alive` `x` `expected number of purchases` gives you something close to it. In any case, that's something the model should be doing implicitly, though it would be nice to have it somewhere indeed.

**Comment by soliverc** [comment]
>I think you could use the sum of each customer's past purchases as an estimate of their CLV and check if `probability of a customer being alive` `x` `expected number of purchases`

OK.. so the model's CLV includes all past transactions in the calculations? I always thought it was lifetime value from _after_ the max date in the customer matrix. As in, how much more money do we expect to make from this customer. 

Also, do you mean `average transaction value` `x` `expected number of purchases`?

**Comment by psygo** [comment]
> OK.. so the model's CLV includes all past transactions in the calculations? I always thought it was lifetime value from *after* the max date in the customer matrix. As in, how much more money do we expect to make from this customer. 

What matters to the model is its parameters &mdash; which are simply obtained from training; and applying them to training or other datasets is valid &mdash; and the RFM (`x`, `t_x`, `T` and `monetary_value`) parameters of the individual. If you want to use a specific point in time that's inside the training set and predict what the customer is going to spend from that point on, it's perfectly fine, as long as you use its RFM parameters from that instant. That's actually what *Maximum Likelihood Estimation* is doing in the background when it fits the model.

> Also, do you mean `average transaction value` `x` `expected number of purchases`?

You are correct, I forgot to include `average transaction value`. However, I don't know if you would still have to multiply all that by the `probability of a customer being alive` also. I *think* that the BG/NBD model already implicitly includes the probability in the calculations, but the Pareto/NBD doesn't. I think it's best that you check within the library and cross-check with the article(s).

**Comment by soliverc** [comment]
I'm getting a better picture of how the model works. I just have a question about the p-alive plots.

According to my plot below, a supermarket customer almost died twice in June and July, and actually died once in August. Obviously he was not dead and came back a short while later. If you saw that a customer was classified as `p-alive = 0`, knowing that he may just purchase again next week - how would you interpret this? 

If I find 1,000 customers to be 'dead' right now in my records today. Obviously I may spend money on some re-activation campaigns etc. But how can I trust that they are actually 'dead' and that I'm not wasting money?

Is the model being too trigger happy in declaring somebody sa `p-alive=0`, and is there a way to adjust this? 

![image](https://user-images.githubusercontent.com/47222261/68791523-6b3ca980-0641-11ea-8017-7a2aeaa57f67.png)

**Comment by psygo** [comment]
It's a probabilistic model, there is no such thing as 100% sure basically. Your model will make "mistakes", there are many variables that are basically random at heart and no model will be able to take into consideration &mdash; for example, how can a model predict that a certain brand went on a crazy sale all of a sudden?

In any case, you can always have a degree of confidence of your results, e.g. I'm 95% sure that, on average, my customers will follow this model. How to calculate this percentage more specifically, I don't know, because it also depends on your metric for it. But it is surely related to the log-likelihood calculated in the background and some of it is exposed in the table shown by the `model.summary()` property.

**Comment by soliverc** [comment]
I totally understand that we can never get 100% certainty. I'd be happy with 75%. I'm just trying to figure out how to answer the question _"So you would have declared my active customer dead almost three times?!!"_ when I present lifetimes as a possible tool to use in work. 

Obviously from the plot above, I would imagine that this customer went on a couple of holidays. It's the height of holiday season in my country. But we can't declare customers as almost dead when they go on holidays and expend €€€ on marketing campaigns to win them back.

It would be useful to have a method to train, test split the data and check the accuracy of p-alive. Similar to the calibration and holdout plot for purchase frequency, I will try and make something myself to test before I present. For example, the model predicted that John Smith died in May. The training cut-off is June. Does John Smith have any purchases after June? If he does, then the model is incorrect etc. etc and so on for however many thousands of customers in the dataset. 

How would you test the accuracy of p-alive? Do you have any advice?

**Comment by VinceZJ** [comment]
> For example, the model predicted that John Smith died in May. The training cut-off is June. Does John Smith have any purchases after June? If he does, then the model is incorrect etc. etc and so on for however many thousands of customers in the dataset.

It seems to me you are coming at this problem from a common supervised learning angle, where we can clearly observe all types of outcomes.
However, by definition of the problem we are trying to solve with the models implemented here, we have some type of non-contractual churn situation, i.e. **no observed churn event**.

This means, while you will be able to figure out the # of true positives (predicted to be alive, and was) & false negatives (predicted not to be alive, but was) with your procedure described above, **there is no way for you to gain any *actual* reliable measurements of false positives & true negatives**

Therefore any type of error-metric you define based on these, will be based on some arbitrary cut-off point/duration (in your case after June until ...?), some arbitrary or weighted p(al8ve)-value, and will ultimately lead to misestimates, since **by the problem definition itself it is not possible to truly classify all observations.**

The difficulty in observing any *actual* churn event in the scenarios also described in the docs, is the reason why these types of probabilistic models were developed in the first place.
